Figaro/src/main/scala/com/cra/figaro/experimental/structured/ProblemComponent.scala

```diff
@@ -179,17 +178,79 @@ extends ExpandableComponent[ParentValue, Value](problem, chain.parent, chain) {
     subproblems += parentValue -> subproblem
   }
 
+  /*
+   * If all the subproblems have been eliminated completely, i.e., the factors returned for each
+   * subproblem mention only the chain outcome, we can use the compact chain factor construction.
+   * This is the case if none of these factors contains variables whose component is not one of the subproblem's
+   * components.
+   */
+  private def allSubproblemsEliminatedCompletely: Boolean = {
+for { (parentValue, subproblem) <- subproblems } {
+  println("parentValue = " + parentValue)
+  println("subproblem = " + subproblem)
+  println("factors = \n" + subproblem.solution.map(_.toReadableString).mkString("\n"))
+}
+    for {
+      (parentValue, subproblem) <- subproblems
+      factor <- subproblem.solution
+      variable <- factor.variables
+    } {
+
+      if (!subproblem.components.contains(problem.collection.variableToComponent(variable))) return false
+    }
+    return true
+  }
+
+  def compactChainFactor: Factor[Double] = {
```

Comment:
This method is kind of long and a little confusing. A comment at the top explaining what the steps are would be helpful. Then, you could make a few comments in the body of the method matching the sections of code to the steps in the outline.